src/compiler/checker.ts

```diff
@@ -31399,6 +31404,70 @@ namespace ts {
+            const exprType = node.kind === SyntaxKind.ExpressionWithTypeArguments ? checkExpression(node.expression) :
+                isThisIdentifier(node.exprName) ? checkThisExpression(node.exprName) :
+                checkExpression(node.exprName);
+            const typeArguments = node.typeArguments;
+            if (!some(typeArguments)) {
+                return exprType;
+            }
+            const funcType = getApparentType(exprType);
+            if (funcType === silentNeverType || isErrorType(funcType)) {
+                return funcType;
+            }
+            let hasSomeApplicableSignature = false;
+            let nonApplicableType: Type | undefined;
+            let hasSignatures: boolean;
+            let hasApplicableSignature: boolean;
+            const result = mapType(funcType, getInstantiatedType);
+            const errorType = hasSomeApplicableSignature ? nonApplicableType : funcType;
+            if (errorType) {
+                diagnostics.add(createDiagnosticForNodeArray(getSourceFileOfNode(node), typeArguments, Diagnostics.Type_0_has_no_signatures_for_which_the_type_argument_list_is_applicable, typeToString(errorType)));
+            }
+            return result;
+
+            function getInstantiatedType(type: Type): Type {
+                hasSignatures = false;
+                hasApplicableSignature = false;
+                const result = getInstantiatedTypeWorker(type);
+                hasSomeApplicableSignature ||= hasApplicableSignature;
+                if (hasSignatures && !hasApplicableSignature) {
+                    nonApplicableType ??= type;
+                }
+                return result;
+            }
+
+            function getInstantiatedTypeWorker(type: Type): Type {
+                if (type.flags & TypeFlags.Object) {
+                    const resolved = resolveStructuredTypeMembers(type as ObjectType);
+                    const callSignatures = getInstantiatedSignatures(resolved.callSignatures);
+                    const constructSignatures = getInstantiatedSignatures(resolved.constructSignatures);
+                    hasSignatures ||= resolved.callSignatures.length !== 0 || resolved.constructSignatures.length !== 0;
+                    hasApplicableSignature ||= callSignatures.length !== 0 || constructSignatures.length !== 0;
+                    if (callSignatures !== resolved.callSignatures || constructSignatures !== resolved.constructSignatures) {
+                        const result = createAnonymousType(undefined, resolved.members, callSignatures, constructSignatures, resolved.indexInfos) as ResolvedType & InstantiationExpressionType;
+                        result.objectFlags |= ObjectFlags.InstantiationExpressionType;
+                        result.node = node;
+                        return result;
+                    }
+                }
+                else if (type.flags & TypeFlags.Intersection) {

```

Comment:
Instantiable types (including substitution types) with constraint types that have applicable signatures are replaced with instantiations of those constraint types. Otherwise instantiable types are simply left unchanged.